tools/firmware-utils/src/tplink-safeloader.c

```diff
@@ -539,6 +539,42 @@ static struct device_info boards[] = {
 		.first_sysupgrade_partition = "os-image",
 		.last_sysupgrade_partition = "support-list",
 	},
+	/** Firmware layout for the CPE710 V1 */
+	{
+		.id     = "CPE710V1",
+		.vendor = "CPE710(TP-LINK|UN|AC866-5|00000000):1.0\r\n",
+		.support_list =
+			"SupportList:\r\n"
+			"CPE710(TP-LINK|UN|AC866-5|00000000):1.0\r\n"
+			"CPE710(TP-LINK|EU|AC866-5|45550000):1.0\r\n"
+			"CPE710(TP-LINK|US|AC866-5|55530000):1.0\r\n"
+			"CPE710(TP-LINK|UN|AC866-5):1.0\r\n"
+			"CPE710(TP-LINK|EU|AC866-5):1.0\r\n"
+			"CPE710(TP-LINK|US|AC866-5):1.0\r\n",
+		.part_trail = 0xff,
+		.soft_ver = NULL,
+
+		.partitions = {
+			{"fs-uboot", 0x00000, 0x50000},
+			{"partition-table", 0x50000, 0x02000},
+			{"default-mac", 0x60000, 0x00020},
+			{"product-info", 0x61100, 0x00100},
+			{"device-info", 0x61400, 0x00400},
+			{"signature", 0x62000, 0x00400},
+			{"device-id", 0x63000, 0x00100},
+			{"firmware", 0x70000, 0xf40000},
+			{"soft-version", 0xfb0000, 0x00100},
+			{"support-list", 0xfb1000, 0x01000},
+			{"user-config", 0xfc0000, 0x10000},
+			{"default-config", 0xfd0000, 0x10000},
+			{"log", 0xfe0000, 0x10000},
+			{"radio", 0xff0000, 0x10000},
+			{NULL, 0, 0}
+		},
```

Comment:
`serial-number` is missing from partitions

```bash
grep -a "partition .*$" CPE710\(UN\)_1.0-up-ver2-2-0-P1\[20200304-rel70345\].bin
partition fs-uboot base 0x00000 size 0x50000
partition partition-table base 0x50000 size 0x02000
partition default-mac base 0x60000 size 0x00020
partition serial-number base 0x60100 size 0x00020
partition product-info base 0x61100 size 0x00100
partition device-info base 0x61400 size 0x00400
partition signature base 0x62000 size 0x00400
partition device-id base 0x63000 size 0x00100
partition os-image base 0x70000 size 0x170000
partition soft-version base 0x1e0000 size 0x00100
partition support-list base 0x1e1000 size 0x01000
partition file-system base 0x1f0000 size 0xdd0000
partition user-config base 0xfc0000 size 0x10000
partition default-config base 0xfd0000 size 0x10000
partition log base 0xfe0000 size 0x10000
partition radio base 0xff0000 size 0x10000
```

Reply:
@jwmullally is you have a look at tools/firmware-utils/src/tplink-safeloader.c
You will find that it is omitted on all the CPE and WBS Devices.
The standard has been to Not include it in the definitions.

Reply:
Without `serial-number` defined, it looks like that data is being erased. /dev/mtd3 offset 0x100 shows as blank. But I can't confirm there was data there in the first place as I've now flashed my 2 units :laughing: 
```
root@OpenWrt:~# cat /proc/mtd 
dev:    size   erasesize  name
mtd0: 00040000 00010000 "u-boot"
mtd1: 00010000 00010000 "u-boot-env"
mtd2: 00010000 00010000 "partition-table"
mtd3: 00010000 00010000 "info"
mtd4: 00f50000 00010000 "firmware"
mtd5: 00220000 00010000 "kernel"
mtd6: 00d30000 00010000 "rootfs"
mtd7: 00990000 00010000 "rootfs_data"
mtd8: 00030000 00010000 "config"
mtd9: 00010000 00010000 "art"
root@OpenWrt:~# hexdump -C /dev/mtd3
00000000  00 00 00 06 00 00 00 00  00 5f 67 03 3f 5c ff ff  |........._g.?\..|
00000010  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00001100  00 00 00 40 00 00 00 00  43 50 45 37 31 30 28 54  |...@....CPE710(T|
00001110  50 2d 4c 49 4e 4b 7c 55  4e 7c 41 43 38 36 36 2d  |P-LINK|UN|AC866-|
00001120  35 7c 30 30 30 30 30 30  30 30 29 3a 31 2e 30 00  |5|00000000):1.0.|
00001130  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00001140  00 00 00 00 00 00 00 00  ff ff ff ff ff ff ff ff  |................|
00001150  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
```

I suspect the TP-Link OEM flash process clears any undefined or misdefined areas ([one example](https://github.com/openwrt/openwrt/pull/2991#discussion_r444491645)), but I don't know for sure.

Reply:
Here is what mine looked like before flashing it for the first time
<pre>*
00011100  00 00 00 28 00 00 00 00  43 50 45 37 31 30 28 54  |...(....CPE710(T|
00011110  50 2d 4c 49 4e 4b 7c 55  53 7c 41 43 38 36 36 2d  |P-LINK|US|AC866-|
00011120  35 7c 35 35 35 33 30 30  30 30 29 3a 31 2e 30 0a  |5|55530000):1.0.|
00011130  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00011140  00 00 00 00 00 00 00 00  ff ff ff ff ff ff ff ff  |................|
00011150  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
</pre>
And Here is the Hardware ID
<pre>00011400  00 00 03 f8 00 00 00 00  68 77 5f 69 64 3a 39 33  |........hw_id:93|
00011410  42 31 42 36 34 46 46 30  36 31 35 38 36 36 30 34  |B1B64FF061586604|
00011420  35 46 42 44 43 39 45 42  38 34 42 32 39 36 0a 6f  |5FBDC9EB84B296.o|
00011430  65 6d 5f 69 64 3a 39 32  33 39 31 43 33 42 30 46  |em_id:92391C3B0F|
00011440  32 33 37 43 46 37 45 42  35 37 30 37 46 32 35 35  |237CF7EB5707F255|
00011450  30 33 46 32 41 44 0a 70  72 6f 64 75 63 74 5f 6e  |03F2AD.product_n|
00011460  61 6d 65 3a 63 70 65 37  31 30 0a 70 72 6f 64 75  |ame:cpe710.produ|
00011470  63 74 5f 76 65 72 3a 31  2e 30 0a 00 00 00 00 00  |ct_ver:1.0......|
00011480  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
</pre>

Reply:
Is there anything at absolute address 0x60100? (which looks like relative addr 0x10100 in your pasted info)?

Reply:
I just tested this by writing data to 0x60100 and flashing factory.bin with and without this line. Without it, this area is cleared during flashing.
```
{"serial-number", 0x60100, 0x00020},
```

Reply:
I don't know about newer revisions, but the CPE210/510 1.0 that I added to OpenWrt never had a serial-number partition. The partition layout for new devices should based on TP-Link's layout for the specific device, and I agree that the serial-number partition must be preserved on CPE710.

Reply:
@NeoRaider @adschm @mpratt14 @jwmullally @robimarko  I believe that all the Newer TP Link CPE/WBS devices have the Serial Number Defined in the vendor firmware but do not define it in the OpenWrt Software.
See Vendor Layout for the TP LINK CPE210 V2 below and was initially added by @robimarko and others
<pre>(CPE210(TP-LINK|UN|N300-2|00000000):2.0
fwup-ptn partition-table base 0x00800 size 0x00800      
fwup-ptn os-image base 0x01000 size 0x100001    
fwup-ptn soft-version base 0x101001 size 0x00015        
fwup-ptn support-list base 0x101016 size 0x000cd        
fwup-ptn file-system base 0x1010e3 size 0x49a001        
partition fs-uboot base 0x00000 size 0x20000
partition partition-table base 0x20000 size 0x02000
partition default-mac base 0x30000 size 0x00020
partition serial-number base 0x30100 size 0x00020
partition product-info base 0x31100 size 0x00100
partition device-info base 0x31400 size 0x00400
partition signature base 0x32000 size 0x00400
partition device-id base 0x33000 size 0x00100
partition os-image base 0x40000 size 0x170000
partition soft-version base 0x1b0000 size 0x00100
partition support-list base 0x1b1000 size 0x01000
partition file-system base 0x1c0000 size 0x600000
partition user-config base 0x7c0000 size 0x10000
partition default-config base 0x7d0000 size 0x10000
partition log base 0x7e0000 size 0x10000
partition radio base 0x7f0000 size 0x10000
</pre>
If you feel it should be defined on the CPE710 then I believe that someone needs to open a PR to address this on ALL the TP-LINK CPE/WBS Devices that have it defined so that they are consistent and then I will update this PR to match as well.

Reply:
The easiest way to ensure correct coverage there is to:

* Download all known CPE firmwares (while being friendly to the HTTP servers)
* Extract the partitions list
* Get all make/models with `serial-number` defined

(Similar to https://github.com/jwmullally/openwrt_stuff/blob/master/tl-wpa8630-v2/report/report.tsv but with different columns).

That should produce a clear list of devices to update.

(You could also check for all partitions, but it seems like this is the only one missing).

If i have time over the next week I can produce that report and/or PR.

Reply:
@jwmullally I have added serial-number for the TP-LINK CPE710. Please test again with your device.

Reply:
Ok, looks like that fixed the serial issue. Tested OEM -> factory.bin and sysupgrades.